Geo/11721 fix settlement output

rfaasse opened the pull request:
**📝 Description**
This PR aims to:
1. Fix that the results after applying a normal load were different from the results when running the settlement through the python interface
2. The results after applying excavation now write 0s for displacements on excavated nodes (like in the python workflow)
3. The settlement outputs for all time steps are now written to file, instead of only the last one.

WPK4FEM committed:
WPK first attempt time loop executor

WPK4FEM committed:
also complementary to WPK first attempt time loop executor

WPK4FEM committed:
Removed yet unused things. Added first unit test: TimeLoopReturnsPerformedStatesAfterRunning. Removed template argument from TimeLoopExecutor class.

WPK4FEM committed:
Added a StrategyWrapper with an always converging solver strategy for testing

WPK4FEM committed:
Added constructor with convergence argument to DummySolverStrategy. Added test for running and terminating a loop with a non-converging solver strategy.

WPK4FEM committed:
Give time loop executor a TimeStepEndState to start from. Keep track of the time passed. Added a test on end time after N steps.

WPK4FEM committed:
Use the time step executor. ( only one available so default constructed ). Passed solver strategy to time step executor.

WPK4FEM committed:
Keep track of used number of cycles. Added tests using a test time incrementor with a fixed number of cycles per step. Added RunCycle and RunCycleLoop to the time loop executor.

avdg81 committed:
Removed a few unused dummy classes for testing

Also added the `explicit` keyword to a single-argument constructor.

avdg81 committed:
Test the time loop executor using a prescribed time incrementor

Given a prescribed time incrementor that keeps four time increments and
a solver strategy that always converges, we expect to get four converged
time step end states, where each one just required a single cycle. Also
the time values are validated for each time step end state.

Furthermore added a helper function to build a converged end state.

avdg81 committed:
Added the strategy wrapper as a member of the time loop executor

Now that the strategy wrapper is a member of the time loop executor, we
can set the end time and the time increment in their "global" variables.
Internally, this is done via the model part. The reason we set these
variables explicitly is that processes may rely on having this
information available.

avdg81 committed:
When running the time loop, increment the step number

Extended the API of class `StrategyWrapper` such that we can increment
the step number and get the step number upon request.

WPK4FEM committed:
Added test for and implementation of CloneTimeStep, in the strategy wrapper needed for the time loop executor.

rfaasse committed:
Merge branch 'master' into settlement_test

WPK4FEM committed:
Added method and tests for RestorePositionsAndDOFVectorToStartOfStep for restoring to the initial situation of a step to enable cycling of the step.

rfaasse committed:
First implementation of the strategy wrapper implementation

rfaasse committed:
Added unit tests for all non-trivial functionality of the wrapper

rfaasse committed:
Fix pipeline build

WPK4FEM committed:
SaveTotalDisplacementFieldAtStartOfStage and AccumulateTotalDisplacementField implemented and unit tests provided.

WPK4FEM committed:
OutputProcess implementation and unit test added to time_loop_executor.

WPK4FEM committed:
FinalizeSolutionStep should be called outside the cycle loop. This changes a test for the TimeStepExecutor and there is a new one for the TimeLoopExecutor.

WPK4FEM committed:
Small things from review by Richard adapted.

WPK4FEM committed:
Pass on process observables from TimeLoopExecutor to TimeStepExecutor.

WPK4FEM committed:
Less code repetition in test_time_loop_executor

WPK4FEM committed:
interface class for TimeLoopExecutor as requested in review by Richard.

WPK4FEM committed:
merge from master

rfaasse committed:
Added all the new strategy wrapper functions

WPK4FEM committed:
forgotten interface.h

WPK4FEM committed:
Code smell remarks from SonarCloud almost all resolved

rfaasse committed:
Added UTs for Getters, setters and Save and Accumulate

rfaasse committed:
Added output process to strategy wrapper

rfaasse committed:
Added functionality for restore positions and dof vector

WPK4FEM committed:
Implemented suggestions from review by Anne.

WPK4FEM committed:
Last? SonarCloud code smell fixed

rfaasse committed:
Added unit test Restoring DoFs

rfaasse committed:
Merge branch 'geo/time-step-loop' into geo/strategy_wrapper_implementation

rfaasse committed:
Merge time loop executor work into feature branch

rfaasse committed:
Added test

rfaasse committed:
Merge branch 'settlement_test' into geo/strategy_wrapper_implementation

rfaasse committed:
Added integration tests

avdg81 committed:
Several small improvements

- No longer track the log file that is created by running the unit test.
- Removed an unused `#include`.
- Use a `std::filesystem::path` object to represent the working
  directory. Also renamed this variable to comply with the Kratos Style
  Guide.
- Corrected a mistake in the names of the project parameter files (the
  word "stage" was missing).
- Open the log file that is created by the unit test only once. And use
  write mode rather than append mode.

avdg81 committed:
Extracted copying of nodal solution step values

When the solving strategy wrapper needs to restore the nodal positions
and the nodal solution step values, it explicitly checks for each nodal
solution step variable whether the model part actually has it. If yes,
then it copies for each node the value at the given source index to the
given destination index.

Split an existing unit test into two new tests. The first one verifies
whether the combination of displacements and water pressures can be
restored. The second one verifies whether the combination of
displacements and rotations can be restored.

avdg81 committed:
Refactored and improved some unit tests

- Extracted a helper function that creates a dummy model part.
- The helper function that creates a solving strategy wrapper now no
  longer creates a dummy model part. Rather it receives one through its
  parameter list.
- Gave some test values a name.
- Corrected some usages of `KRATOS_EXPECT_EQ` to
  `KRATOS_EXPECT_DOUBLE_EQ`.

avdg81 committed:
Refactored and improved some unit tests

- The helper function that creates a solving strategy wrapper with
  default process info entries now no longer creates a dummy model part.
  Rather it receives one through its parameter list.
- Eliminated a few unneeded local variables.
- The test for incrementing the step number now increments and verifies
  the step number twice.
- Corrected a few variable names to comply with the style guide.

avdg81 committed:
Added the adaptive time incrementor to the settlement workflow

- Added several helper functions to extract the parameters for the
  adaptive time incrementor from the project parameters.
- The time loop executor of the settlement workflow now receives an
  adaptive time incrementor.
- The factory that creates the settlement workflow now uses a valid time
  loop executor.

carloslubbers committed:
Merge branch 'master' into geo/strategy_wrapper_implementation

# Conflicts:
#	applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.cpp

carloslubbers committed:
Add logging when new time step is executed

avdg81 committed:
Added a working solving strategy wrapper to the settlement workflow

- Create a solving strategy wrapper instance and pass it to the time
  loop executor of the settlement workflow.
- The solving strategy factory now accepts "newton_raphson" as strategy
  type rather than "newton_raphson_strategy" (which is the name returned
  by that strategy). This is in line with the Python workflow of the
  settlement calculation.

rfaasse committed:
Added linear solvers application link to settlement and geomechanics

avdg81 committed:
The settlement workflow now tries to run the time loop executor

For now, we pass a dummy time step end state to the `Run` member
function.

Note that the unit test of the settlement workflow now ends with a
crash!

avdg81 committed:
Corrected a mistake when creating the linear solvers application

- Corrected the type of data member `mpLinearSolverApp`. It used to be
  `KratosGeoMechanicsApplication::Pointer`, whereas
  `KratosLinearSolversApplication::Pointer` was intended.
- Also corrected the statement that creates this smart pointer.
- Use the plural form (linear solvers) in the name of the data member,
  to be consistent with the name of the application.

avdg81 committed:
Add degrees of freedom to after reading the `.mdpa` file

This is in line with what the D-GeoFlow workflow does and it seems to
resolve a crash. (Carlos had discovered the same thing.)

avdg81 committed:
Print lots of debug messages and fixed a few minor issues

- Implemented the `Info` member functions of classes
  `ApplyK0ProcedureProcess`, `ApplyScalarConstraintTableProcess`,
  `ApplyVectorConstraintTableProcess`, and
  `BackwardEulerQuasistaticUPwScheme`.
- Explicitly set the number of threads to be used by the settlement
  workflow to 1 (in line with what is done by the geoflow workflow).
- Included the fix for the model parts used by the process creators for
  the settlement workflow.
- Properly initialize the processes used by the settlement workflow.
- Added a new model part that is referenced by the K0 procedure.
- Added the K0 procedure to the list of processes for the first stage of
  the example settlement workflow.
- Added a few checks for "empty" objects when creating a solver
  strategy.
- For testing purposes, we have removed the conditions entirely from the
  `.mdpa` file of the example settlement analysis. Furthermore, we have
  removed all the submodel parts that are not used during the first
  stage of the example settlement analysis.

rfaasse committed:
Added missing gauss parameters to writing the output. The settlement workflow test now passes for stage 1

avdg81 committed:
Restored the conditions and several submodel parts

For testing purposes, those had been temporarily removed. Since we now
handle the model parts given to the processes in a better way, we should
try to restore those.

rfaasse committed:
Made fix for option to omit hydraulic head at nodes and removed log file for test to fix pipeline. First three stages now complete without crashing

rfaasse committed:
Added python res files

avdg81 committed:
Create a computational model part

In line with the approach followed by the Python scripts, we need to
have a computational (sub)model part that keeps "active" parts of the
mesh/model, including the nodes, elements, and conditions of the main
model part. It also shares the process info with the main model part.
This computational model part is then passed to the factory that creates
solving strategies.

When the solving strategy wrapper attempts to clone a time step, we now
either pass its model part (if it is not a submodel part) or the root
model part of its (sub)model part.

Furthermore, we added a few more print statements for debugging.

rfaasse committed:
Removed logging from core library

avdg81 committed:
Fixed the times written to the GiD output files

Also accepted changed results for the D-GeoFlow "unit tests".

rfaasse committed:
reverting core files to master state

avdg81 committed:
Fixed several code smells reported by SonarQube

- Replaced a deprecated function call by its successor.
- Changed the types of two variables to a reference-to-const.
- Changed the type of a parameter to a reference-to-const.

rfaasse committed:
Fix GCC build

rfaasse committed:
Removed unnecessary logging

rfaasse committed:
Extracted duplicate code to fix build

rfaasse committed:
Merge branch 'master' into geo/strategy_wrapper_implementation

rfaasse committed:
Small fix after merge

WPK4FEM committed:
Introduced the spatial tables of Marjan into the settlement test model

WPK4FEM committed:
Merge remote-tracking branch 'refs/remotes/origin/geo/strategy_wrapper_implementation' into geo/strategy_wrapper_implementation

rfaasse committed:
Reduced code smells

WPK4FEM committed:
GeoElementUtilities::CalculateNodalHydraulicHeadFromWaterPressures now gets its size for the returned vector from the number of nodes of the element.

WPK4FEM committed:
Merge remote-tracking branch 'refs/remotes/origin/geo/strategy_wrapper_implementation' into geo/strategy_wrapper_implementation

rfaasse committed:
Remove redundant print statement

avdg81 committed:
Merge branch 'geo/strategy_wrapper_implementation' of github.com:KratosMultiphysics/Kratos into geo/strategy_wrapper_implementation

avdg81 committed:
Simplified the writing of integration point (and nodal) results

We no longer need to cast and test a variable to determine which
function overload we need for writing the variable's results. We now
build up a map of integration point (and nodal) result writers, which in
turn are created by two factory function templates. Adding a new result
to be output is a matter of adding a single line.

rfaasse committed:
Added copying properties

avdg81 committed:
Removed a redundant blank line

rfaasse committed:
removed redundant log file

rfaasse committed:
Cleaning up

rfaasse committed:
Merge branch 'geo/strategy_wrapper_implementation' of https://github.com/KratosMultiphysics/Kratos into geo/strategy_wrapper_implementation

rfaasse committed:
Moved settlement workflow test to integration suite of cpp tests

avdg81 committed:
Reordered a few lines to make the flow more logical

rfaasse committed:
Removing SQ issues

rfaasse committed:
Merge remote-tracking branch 'origin/geo/strategy_wrapper_implementation' into geo/strategy_wrapper_implementation

avdg81 committed:
Two minor enhancements

1. Prefer to use the pre-increment operator rather than the
   post-increment operator.
2. Made a local variable `const`.

avdg81 committed:
Removed several log statements

avdg81 committed:
Removed a redundant statement

A node that has been created by a model part does not need to be added
explicitly to that same model part later on.

avdg81 committed:
Removed a redundant destructor

rfaasse committed:
Added test for saving total replacements twice and other minor fixes

rfaasse committed:
Fix Codacy complaint

rfaasse committed:
Fix UT

rfaasse committed:
Only run stage 1-3 in the workflow test, since stage 4 has known issues

rfaasse committed:
Merge branch 'master' into geo/strategy_wrapper_implementation

rfaasse committed:
Possible fix for pipeline test failure

rfaasse committed:
Another attempt at fixing the pipeline test

rfaasse committed:
Skipping workflow test temporarily due to discussed pipeline issues

rfaasse committed:
Fixing normal load

rfaasse committed:
Merge branch 'master' into geo/11721-fix-settlement-output

rfaasse committed:
Minor change to test and fixed the displacement for excavated part of model

rfaasse committed:
Fixed writing multiple time steps in one file

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.cpp:
```diff
@@ -231,9 +234,14 @@ int KratosGeoSettlement::RunStage(const std::filesystem::path&            rWorki
-            mpTimeLoopExecutor->Run(dummy_state);
+            TimeStepEndState start_of_loop_state;
+            start_of_loop_state.convergence_state = TimeStepEndState::ConvergenceState::converged;
+            start_of_loop_state.time = GetStartTimeFrom(project_parameters);
```
Without this line, the loop would always start at 0 (even if the start time e.g. of stage 4 is 3). Since we only wrote the output for the last time step (it would overwrite the .res file after every time step), we didn't see it before.

avdg81 replied:
Good catch! Thanks for fixing this issue.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.cpp:
```diff
@@ -353,6 +361,31 @@ std::shared_ptr<StrategyWrapper> KratosGeoSettlement::MakeStrategyWrapper(const
+//            node.SetSolutionStepValue(KratosMultiphysics.ROTATION, 1, rotNew);
+        }
+
+    }
```
This is one of the parts we missed and this fixes writing the 0s on the excavated nodes. You can find it in geomechanics_analysis.py line 68-78. I didn't have time to clean it up (ROTATION is commented out because it doesn't exist in our example). I was thinking to do it in the same way as we do it in the solving_strategy_wrapper (see the CopyNodalSolutionStepValues function), but **note** that the modelpart used here is the computational model part, not the main model part

avdg81 replied:
Yes, you are right. I have tried to simplify the code a bit in the way you suggested. Furthermore, I've noticed that there was apparently one more statement missing (see line 78 of file `geomechanics_analysis.py`). I've added it to our C++ code (see line 371 of file `dgeosettlement.cpp`).

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.cpp:
```diff
@@ -353,6 +361,31 @@ std::shared_ptr<StrategyWrapper> KratosGeoSettlement::MakeStrategyWrapper(const
+    find_neighbours_process.Execute();
+
+    auto deactivate_conditions_on_inactive_elements_process = DeactivateConditionsOnInactiveElements(main_model_part.GetSubModelPart(mComputationalSubModelPartName));
+    deactivate_conditions_on_inactive_elements_process.Execute();
```
This part (which really needs to be done **after** solving strategy initialization), fixed the differences we saw for the normal loads. If the part above is moved to the solving_strategy_wrapper, this might also have to be moved

avdg81 replied:
Good catch.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/solving_strategy_wrapper.hpp:
```diff
@@ -180,6 +192,7 @@ class SolvingStrategyWrapper : public StrategyWrapper
     std::vector<array_1d<double, 3>> mOldTotalDisplacements;
     Parameters mProjectParameters;
     std::filesystem::path mWorkingDirectory;
+    std::unique_ptr<GeoOutputWriter> mWriter;
```
I had to cache the writer, so it wouldn't reset after each time step

avdg81 replied:
Good catch again.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/solving_strategy_wrapper.hpp:
```diff
@@ -12,39 +12,39 @@
 
```
Used clang format for this file, sorry that makes the reviewing a bit harder

avdg81 replied:
No problem. We will start to use clang-format anyway at some point in time :-)

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/time_loop_executor.hpp:
```diff
@@ -48,6 +48,7 @@ public :
 
     std::vector<TimeStepEndState> Run(const TimeStepEndState& EndState) override
     {
+        mStrategyWrapper->Initialize();
```
Moved this Initialize from the time step executor to the time loop executor, since that was the same in python (and it didn't break anything)

avdg81 replied:
Well done.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/write_output.cpp:
```diff
@@ -193,4 +179,24 @@ WriteConditionsFlag GeoOutputWriter::GetWriteConditionsFlagFrom(const Parameters
+        GetGiDPostModeFrom(Settings["output_processes"]["gid_output"].GetArrayItem(0)["Parameters"]["postprocess_parameters"]["result_file_configuration"]["gidpost_flags"]),
+        GetMultiFileFlagFrom(Settings["output_processes"]["gid_output"].GetArrayItem(0)["Parameters"]["postprocess_parameters"]["result_file_configuration"]["gidpost_flags"]),
+        GetWriteDeformedMeshFlagFrom(Settings["output_processes"]["gid_output"].GetArrayItem(0)["Parameters"]["postprocess_parameters"]["result_file_configuration"]["gidpost_flags"]),
+        GetWriteConditionsFlagFrom(Settings["output_processes"]["gid_output"].GetArrayItem(0)["Parameters"]["postprocess_parameters"]["result_file_configuration"]["gidpost_flags"])}
```
Yes, this is very messy, I didn't have time to fix it unfortunately.

avdg81 replied:
I have taken care of that.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/write_output.cpp:
```diff
@@ -193,4 +179,24 @@ WriteConditionsFlag GeoOutputWriter::GetWriteConditionsFlagFrom(const Parameters
+    mGidIO.InitializeMesh(0.0);
+    mGidIO.WriteMesh(rModelPart.GetMesh());
+    mGidIO.FinalizeMesh();
+    mGidIO.InitializeResults(0, rModelPart.GetMesh());
```
Only done once per stage

avdg81 replied:
It's OK for now, but I think we may want to revisit it, since the initialization and finalization are not symmetric.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/write_output.cpp:
```diff
@@ -193,4 +179,24 @@ WriteConditionsFlag GeoOutputWriter::GetWriteConditionsFlagFrom(const Parameters
+
+void GeoOutputWriter::FinalizeResults()
+{
+    mGidIO.FinalizeResults();
```
Only done once per stage, so the file isn't closed in between the time steps.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_settlement_workflow.cpp:
```diff
@@ -26,15 +26,18 @@ using namespace Kratos;
 {
 
-[[maybe_unused]] void TestSettlement()
+KRATOS_TEST_CASE_IN_SUITE(SettlementWorkflow, KratosGeoMechanicsFastSuite)
```
Don't forget to move this back to [[maybe_unused]] void TestSettlement() before merging (okay it will block, so probably you can't forget)

avdg81 replied:
Thanks for the reminder. I have rolled back the changes such that the build checks will pass again.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_settlement_workflow/MaterialParameters.json:
```diff
@@ -13,7 +13,7 @@
               "DENSITY_WATER"            :  1.0,
               "POROSITY"                 :  0.3,
-              "BULK_MODULUS_SOLID"       :  1.0e9,
+              "BULK_MODULUS_SOLID"       :  35.0e6,
```
These are changes proposed by Marjan and Wijtze Pieter for the test case

avdg81 replied:
Looks good to me.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_settlement_workflow/ProjectParameters_stage4.json:
```diff
@@ -19,7 +19,7 @@
         },
         "time_stepping":              {
-            "time_step":                1.0,
+            "time_step":                0.2,
```
For stage 4 we now have multiple steps (3.2, 3.6 and 4.0 so the adaptive scheme is scheming away!)

avdg81 replied:
Nice check. And good to see that the adaptive time incrementor works as intended.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_settlement_workflow/test_model_stage4.post.orig.res:
This is the result as run by python, the C++ version now matches it

avdg81 committed:
Code cleanup and minor corrections

- Prefer to use curly braces when constructing an object.
- Removed a comment that was no longer correct due to changes.
- Moved the definition of `GeoOutputWriter`'s constructor to the top of
  the file.
- Only pass the GiD output settings to `GeoOutputWriter`'s constructor
  and its member function `WriteGiDOutput`, since we only need settings
  from that section of the project parameters. This simplifies the
  implementation of both member functions.
- Removed an unused local variable.
- Removed a redundant local variable.
- Removed an unused function parameter.
- Removed several redundant function parameters.
- Use more compact notation to get an array item.
- Changed some private member functions to local helpers (to clean up
  `GeoOutputWriter`'s interface).
- Pass a function by reference to const rather than by value.
- Extracted a helper function to create a `GidIO<>` object.
- Extracted member functions that return the main model part and the
  computational model part.
- Extracted a member function that restores the values of the given
  nodal variable.
- The `ROTATION` variable is now also restored when it is present.
- Added an apparently missing update to the initial configuration.
- Simplified some pieces of code.
- A few minor style changes.

avdg81 committed:
Updated the settlement workflow "unit test"

- Reverted the changes that actually run the test and where we log to a
  local file. If we don't do that, the build checks on GitHub fail for
  unknown reasons. When we run the test locally or on our own build
  server all is fine.
- Increased the maximum stage number in the test from 3 to 4.

Furthermore, a minor style change has been made, and a path has been
reverted in a Python script.

avdg81 reviewed (approved):
Really nice and clean changes. Thanks Richard!
